### PR TITLE
LibJS: Rewrite Parser.parse_object_expression()

### LINKED ISSUE #2359: LibJS: Object expression parsing is flawed
Especially since getter/setter shorthand syntax was added, it happily accepts a surprising amount of invalid syntax.

Probably easiest to just rewrite all of `Parser::parse_object_expression()`, it's become a bit of a mess anyway.

```
> ({ get ...foo })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          SpreadExpression
            Identifier "foo"
          SpreadExpression
            Identifier "foo"
```

```
> ({ get... foo })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          SpreadExpression
            Identifier "foo"
          SpreadExpression
            Identifier "foo"
```

```
> ({ get foo })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          StringLiteral "foo"
          Identifier "foo"
```

```
> ({ get foo: bar })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          StringLiteral "foo"
          Identifier "bar"
```

```
> ({ get [foo]: bar })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          Identifier "foo"
          Identifier "bar"
```

```
> ({ get ...[foo] })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          SpreadExpression
            ArrayExpression
              Identifier "foo"
          SpreadExpression
            ArrayExpression
              Identifier "foo"
```

```
> ({ get foo(bar, baz) {} })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          StringLiteral "foo"
          FunctionExpression '(null)'
            (Parameters)
              bar
              baz
            (Body)
              BlockStatement
```

```
> ({ set foo(bar, baz) {} })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          StringLiteral "foo"
          FunctionExpression '(null)'
            (Parameters)
              bar
              baz
            (Body)
              BlockStatement
```

```
> ({ set foo() {} })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          StringLiteral "foo"
          FunctionExpression '(null)'
            (Body)
              BlockStatement
```

```
> ({ ...foo: bar })
Program
  (Children)
    ExpressionStatement
      ObjectExpression
        ObjectProperty
          SpreadExpression
            Identifier "foo"
          Identifier "bar"
```

### PR DESCRIPTION
**Pull request by mattco98**
This rewrite drastically increases the accuracy of object literals. Additionally, an `assertIsSyntaxError` function has been added to `test-common.js` to assist in testing syntax errors.

Fixes #2359, and adds all of the cases presented in that issue as tests to make sure they result in syntax errors.

### CODE REVIEW

File: Libraries/LibJS/AST.h [hunk 1]
```diff
@@ -812,7 +816,7 @@ class ObjectProperty final : public ASTNode {
     virtual const char* class_name() const override { return "ObjectProperty"; }
 
     NonnullRefPtr<Expression> m_key;
-    NonnullRefPtr<Expression> m_value;
+    RefPtr<Expression> m_value;
```
Comment (awesomekling) [inline Libraries/LibJS/AST.h:819]: Why this change? My understanding is that `ObjectProperty` must always have a `value`.

File: Libraries/LibJS/Parser.cpp [hunk 2]
```diff
@@ -514,73 +514,115 @@ NonnullRefPtr<Expression> Parser::parse_unary_prefixed_expression()
     }
 }
 
-NonnullRefPtr<ObjectExpression> Parser::parse_object_expression()
+NonnullRefPtr<Expression> Parser::parse_object_expression()
```
[the remark this answers is not in the document]
Comment (mattco98) [reply Libraries/LibJS/Parser.cpp:517]: So it can return an `ErrorExpression`

File: Libraries/LibJS/AST.h (hunk 1 above)
Comment (mattco98) [reply Libraries/LibJS/AST.h:819]: Not with a spread expression, which only has one value: `({ ...foo })`

Comment (linusg) [reply Libraries/LibJS/AST.h:819]: Other parsers have a separate class for that (something like `SpreadProperty`), maybe that would make sense for us too?

Comment (mattco98) [reply Libraries/LibJS/AST.h:819]: Well it's already differentiated with a `Type` value, which is checked when executing the `ObjectExpression`. Seems like it would just add complexity since then it couldn't be a `Vector<ObjectExpression>`, we'd have to make a base class, and the execute method would have to do a bunch of casting. Just my two cents though.

File: Libraries/LibJS/Parser.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunverwerth) [reply Libraries/LibJS/Parser.cpp:517]: I see. You should not return an `ErrorExpression` at that point. Calling `syntax_error` already marks the AST as invalid. This is in line with the other parser methods. `ErrorExpression` is meant as a placeholder for a real expression in cases where an expression is expected per the grammar but no production rule could be chosen.

[the remark this answers is not in the document]
Comment (sunverwerth) [reply Libraries/LibJS/Parser.cpp:517]: You don't have to return at all. You can just continue parsing the rest of the object.

[the remark this answers is not in the document]
Comment (mattco98) [reply Libraries/LibJS/Parser.cpp:517]: Is there a reason to continue? The overall object expression is still going to throw a syntax error. Plus, if the getter doesn’t have a parenthesis, it’ll be in a really strange state and probably generate more syntax errors.

[the remark this answers is not in the document]
Comment (sunverwerth) [reply Libraries/LibJS/Parser.cpp:517]: Hm, I think we had this discussion before... :)

I guess the parser is rather unorthodox. Currently it's going to parse until it runs out of tokens anyway. I did it that way to have possible error recovery later to use it in HackStudio. It's not so useful when used as a runtime parser, I'm working on something to alleviate this.

Right now it's simply easier to pretend everything is fine and continue parsing.

[the remark this answers is not in the document]
Comment (linusg) [reply Libraries/LibJS/Parser.cpp:517]: @mattco98 you might be interested in https://github.com/SerenityOS/serenity/issues/2207 :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (mattco98) [reply Libraries/LibJS/Parser.cpp:517]: Ahh, that makes complete sense! I will do my best to continue parsing then :) thanks guys

[the remark this answers is not in the document]
Comment (sunverwerth) [reply Libraries/LibJS/Parser.cpp:517]: Great work, btw. Just trying to keep this monster of a codebase in check.
